File: DopamiNah/Features/Stats/StatsViewModel.swift
import Foundation
import SwiftUI
import Combine

enum StatsTab: String, CaseIterable {
    case weekly = "Semana"
    case monthly = "Mes"
}

struct StatsState {
    var selectedTab: StatsTab = .weekly
    var dailyAverageText: String = "-"
    var unlockAverageText: String = "-"
    var dailyUsageMinutes: [Float] = []
    var appUsageData: [AppUsageEntry] = []
    var hourlyUsage: [Float] = Array(repeating: 0, count: 24)
    var selectedDayOffset: Int = 0
    var dailyDetails: DailyDetailStats?
    var isLoading: Bool = false
}

@MainActor
final class StatsViewModel: ObservableObject {
    @Published var uiState = StatsState()

    private let deviceUsageRepo: DeviceUsageRepositoryProtocol

    init(deviceUsageRepo: DeviceUsageRepositoryProtocol = ManualDeviceUsageRepository()) {
        self.deviceUsageRepo = deviceUsageRepo
    }

    func loadData() async {
        uiState.isLoading = true

        let days = uiState.selectedTab == .weekly ? 7 : 30

        async let avgUsage = deviceUsageRepo.getAverageUsageMillis(days: days)
        async let avgUnlocks = deviceUsageRepo.getAverageUnlocks(days: days)
        async let dailyData = deviceUsageRepo.getDailyUsageForLastDays(days: days)
        async let appUsage = deviceUsageRepo.getAverageUsagePerApp(days: days, limit: 8)
        async let hourly = deviceUsageRepo.getHourlyUsage(days: days)
        async let details = deviceUsageRepo.getDailyDetails(dayOffset: uiState.selectedDayOffset)

        let avgUsageResult = await avgUsage
        uiState.dailyAverageText = avgUsageResult.formattedUsageTime
        uiState.unlockAverageText = "\(await avgUnlocks)"
        uiState.dailyUsageMinutes = await dailyData.map { Float($0) / 60_000 }
        uiState.appUsageData = await appUsage.map { AppUsageEntry(appName: $0, averageHours: Float($1) / 3_600_000) }
        uiState.hourlyUsage = await hourly
        uiState.dailyDetails = await details

        uiState.isLoading = false
    }

    func selectTab(_ tab: StatsTab) {
        uiState.selectedTab = tab
        uiState.selectedDayOffset = 0
        let days = tab == .weekly ? 7 : 30
        Task {
            uiState.isLoading = true
            async let dailyData = deviceUsageRepo.getDailyUsageForLastDays(days: days)
            async let appUsage = deviceUsageRepo.getAverageUsagePerApp(days: days, limit: 8)
            async let hourly = deviceUsageRepo.getHourlyUsage(days: days)
            async let details = deviceUsageRepo.getDailyDetails(dayOffset: 0)
            async let avgUsage = deviceUsageRepo.getAverageUsageMillis(days: days)
            async let avgUnlocks = deviceUsageRepo.getAverageUnlocks(days: days)

            uiState.dailyUsageMinutes = await dailyData.map { Float($0) / 60_000 }
            uiState.appUsageData = await appUsage.map { AppUsageEntry(appName: $0, averageHours: Float($1) / 3_600_000) }
            uiState.hourlyUsage = await hourly
            uiState.dailyDetails = await details
            uiState.dailyAverageText = (await avgUsage).formattedUsageTime
            uiState.unlockAverageText = "\(await avgUnlocks)"
            uiState.isLoading = false
        }
    }

    func goToPreviousDay() {
        guard uiState.selectedDayOffset < 30 else { return }
        uiState.selectedDayOffset += 1
        Task {
            uiState.dailyDetails = await deviceUsageRepo.getDailyDetails(dayOffset: uiState.selectedDayOffset)
        }
    }

    func goToNextDay() {
        guard uiState.selectedDayOffset > 0 else { return }
        uiState.selectedDayOffset -= 1
        Task {
            uiState.dailyDetails = await deviceUsageRepo.getDailyDetails(dayOffset: uiState.selectedDayOffset)
        }
    }

    func selectDay(_ dayOffset: Int) {
        uiState.selectedDayOffset = dayOffset
        Task {
            uiState.dailyDetails = await deviceUsageRepo.getDailyDetails(dayOffset: dayOffset)
        }
    }
}

File: shared/src/commonMain/kotlin/co/edu/unicauca/dopaminah/ui/screens/stats/viewmodel/StatsViewModel.kt
package co.edu.unicauca.dopaminah.ui.screens.stats.viewmodel

import co.edu.unicauca.dopaminah.domain.repository.DailyDetailStats
import co.edu.unicauca.dopaminah.domain.repository.DeviceUsageRepository
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/** Time range tab for the stats view. */
enum class StatsTab { WEEKLY, MONTHLY }

/** Usage summary for a single app, used in stats charts. */
data class AppUsageEntry(
    val appName: String,
    val averageHours: Float
)

/** UI state for the stats screen, holding all chart and detail data. */
data class StatsState(
    val selectedTab: StatsTab = StatsTab.WEEKLY,
    val dailyAverageText: String = "-",
    val unlockAverageText: String = "-",
    val lastWeekUsage: List<Float> = emptyList(),
    val appUsageData: List<AppUsageEntry> = emptyList(),
    val hourlyUsage: List<Float> = emptyList(),
    val selectedDayOffset: Int = 0,
    val dailyDetails: DailyDetailStats? = null,
    val isLoading: Boolean = false
)

/** ViewModel for the usage stats screen, providing weekly/monthly data, app breakdown, and hourly charts. */
class StatsViewModel(
    private val repository: DeviceUsageRepository? = null
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    private val _uiState = MutableStateFlow(StatsState(isLoading = true))
    val uiState: StateFlow<StatsState> = _uiState.asStateFlow()

    init {
        if (repository != null) loadData()
        else _uiState.value = StatsState(
            dailyAverageText = "3h 12m",
            unlockAverageText = "45/día",
            lastWeekUsage = listOf(2.5f, 3.1f, 4.0f, 2.8f, 3.5f, 4.2f, 3.0f),
            appUsageData = listOf(
                AppUsageEntry("Instagram", 1.5f),
                AppUsageEntry("WhatsApp", 1.2f),
                AppUsageEntry("YouTube", 0.8f),
                AppUsageEntry("TikTok", 0.6f),
            ),
            hourlyUsage = List(24) { i -> if (i in 8..23) (kotlin.math.sin(i.toFloat() / 4f) + 1f) * 15f else 0f },
            dailyDetails = DailyDetailStats("lunes, 24 de marzo", "7:45 AM", 12, "Instagram", "1h 32m", 15, 3600000L)
        )
    }

    private fun loadData() {
        val repo = repository ?: return
        scope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)
            val days = if (_uiState.value.selectedTab == StatsTab.WEEKLY) 7 else 30
            val avgUsage = repo.getAverageUsageMillis(days)
            val avgUnlocks = repo.getAverageUnlocks(days)
            val historyMillis = repo.getDailyUsageForLastDays(7)
            val historyHours = historyMillis.map { it.toFloat() / (1000f * 60f * 60f) }
            val hourlyUsage = repo.getHourlyUsage(days)
            val rawAppData = repo.getAverageUsagePerApp(days)
            val appEntries = rawAppData.map { (name, millis) ->
                AppUsageEntry(
                    appName = name,
                    averageHours = millis.toFloat() / (1000f * 60f * 60f)
                )
            }
            val details = repo.getDailyDetails(_uiState.value.selectedDayOffset)
            _uiState.value = _uiState.value.copy(
                dailyAverageText = formatTime(avgUsage),
                unlockAverageText = "$avgUnlocks/día",
                lastWeekUsage = historyHours,
                appUsageData = appEntries,
                hourlyUsage = hourlyUsage,
                dailyDetails = details,
                isLoading = false
            )
        }
    }

    fun selectTab(tab: StatsTab) {
        if (_uiState.value.selectedTab != tab) {
            _uiState.value = _uiState.value.copy(selectedTab = tab)
            if (repository != null) loadData()
        }
    }

    fun goToPreviousDay() {
        val current = _uiState.value.selectedDayOffset
        if (current < 30) {
            _uiState.value = _uiState.value.copy(selectedDayOffset = current + 1)
            reloadDailyDetails()
        }
    }

    fun goToNextDay() {
        val current = _uiState.value.selectedDayOffset
        if (current > 0) {
            _uiState.value = _uiState.value.copy(selectedDayOffset = current - 1)
            reloadDailyDetails()
        }
    }

    fun selectDay(dayOffset: Int) {
        val clamped = dayOffset.coerceIn(0, 30)
        if (_uiState.value.selectedDayOffset != clamped) {
            _uiState.value = _uiState.value.copy(selectedDayOffset = clamped)
            reloadDailyDetails()
        }
    }

    private fun reloadDailyDetails() {
        scope.launch {
            if (repository != null) {
                val details = repository.getDailyDetails(_uiState.value.selectedDayOffset)
                _uiState.value = _uiState.value.copy(dailyDetails = details)
            }
        }
    }

    private fun formatTime(millis: Long): String {
        val totalMinutes = millis / 1000 / 60
        val hours = totalMinutes / 60
        val minutes = totalMinutes % 60
        return if (hours > 0) "${hours}h ${minutes}m" else "${minutes}m"
    }
}
